### ISSUE TITLE
[circle-resizer] Implementation a new tool for model resizing

### ISSUE THREAD

**Issue by mbencer** [Samsung/ONE#14791]
In the current moment we have no a capabilities for model resizing and dynamic input specialization.
The best place in the whole tool-chain seems to be operating on circle representation (common solution for all front-ends).

### What types of API should be exposed?
We need both C++ and Python API. Python API is especially needed for Python-based frontends, like Pytorch->Circle converter.
There are also some C++ tools run from Python as sub-processes but I am against such approach (very hard to debug, need to determine paths for every usage, no versioning control, each step requires saving the whole model to file).

### Why a separate tool instead of extending existing API:
I am just follow convention from https://github.com/Samsung/ONE/tree/master/compiler where each step is realized as a separate app/service. What's more the tool should be easy accessible for front-ends.

### Initial use-cases:
In the beginning I am focused on LLM Pytoch models. My first candidate to provide e2e support LLama2. 

### cmd app
The tool should be accessible also as command line app with args:
- `--input_model`
- `--output_model`
- `--input_shapes`
Supported format of new inputs shapes: comma separated. An example for model with 2 inputs: [1,2],[3,4].

### Way of tests:
- unit tests (c++ and python parts)
- e2e for the models, like: Pytorch->Circle conversion to dynamic version, resizing/shape specialization using `circle-resizer`, numerical results testing on [circle-interpreter](https://github.sec.samsung.net/AIP/python-tools/tree/main/pkgs/circle-interpreter)

Proposition of implementation:

![Image](https://github.com/user-attachments/assets/8bb60699-97a0-4d8e-b695-ab224f4dcee8)

### TODO:
- [ ] changing shapes of inputs on `circle::Model` flatbuffer representation
- [ ] shape inference on model with changed shapes using `luci/Pass` capabilities
- [ ] parsing new shapes passed by a user
- [ ] adjust/expand current shape inference capabilities (`luci/Pass/CircleShapeInferencePass`) to support requires cases
- [ ] API to exported changed model to stream/file
- [ ] exposing resizer API using `pybind11`
- [ ] `setup.py` and python wheel packing/distribution

**Comment by mbencer** [comment]
DRAFT: https://github.com/Samsung/ONE/pull/14727

**Comment by seanshpark** [comment]
1/ Can you explain more about what to `resize`?
Providing some examples should help understanding.

2/ Do you have some specific real model that you are interested in? Can you share it?

**Comment by mbencer** [comment]
@seanshpark Sure
> 1/ Can you explain more about what to resize?
> Providing some examples should help understanding.

1. Shape specialization. There are many models with dynamic inputs (inputs which have dynamic shapes) which have to be static before inference (I am not aware of any backend in ONE which accept inputs with dynamic shapes).
2. Input tensor resizing from static-to-static. Like changing batch or input image size in some cases.

Such tool on circle representation level seems to be a good solution for all frontends. Note that some time ago I've proposed sth similar to legacy onnx frontend - https://github.com/Samsung/ONE/pull/13638. In that time it was attempt to resolve a request from our user.

Something to consider - I'm thinking now of changing name of the tool to sth more general - like `circle-editor`. Maybe in the future features like changing data type, layout order, inserting pre-processing layers, cutting some fragments can be also useful.


> 2/ Do you have some specific real model that you are interested in? Can you share it?

My first candidate in `LLama2` with dynamic batch and sequence length. It's proposed in `circle-exir` repo under issue 293 (not sure if I can paste the full link here).

**Comment by seanshpark** [comment]
Thanks for the explanation.

> ... which have to be static before inference

have you checked `circle2circle` tool ? restricted but there is `dynamic_batch_to_single_batch` option

> Maybe in the future features like changing data type, layout order, inserting pre-processing layers, cutting ...

have you checked `circle2circle` tools options? there seems similar options, but I don't know what you are thinking, so, please describe more about the future thoughts if you like, and describe what are the differences.

As I know you are to provide a new tool.
Our individual tools are integrated through `one-cmds`, `onecc` tool and also through sub tools,
if the feature is for users, to manipulate model files.
Please integrate it though this tool, and also prepare some UI tests like others,
though your draft PR, so that reviewers can provide some comments if something goes in different direction.

**Comment by seanshpark** [comment]
> It's proposed in circle-exir repo under issue 293 (not sure if I can paste the full link here).

please preapre a `res/TensorFlowLiteRecipes` recipes with simplified version of the model; each dim < 10 in the shape, so that `luci-interpreter` execution doesn't take too much time.
as I assume, there are lots of Ops in the model, you may have to prepare multiple models,
some for single Ops, and with multiple Ops.

**Comment by mbencer** [comment]
@seanshpark 

> have you checked circle2circle tool ? restricted but there is dynamic_batch_to_single_batch option

Yes. I've seen it but it doesn't covert cases fulfilled by `circle-resizer`.
Do you think that resizing feature such be an option of `circle2circle`? The integration is possible but I've tried to follow rule of rather small services (similar to the rest from https://github.com/Samsung/ONE/tree/master/compiler).

> have you checked circle2circle tools options? there seems similar options, but I don't know what you are thinking, so, please describe more about the future thoughts if you like, and describe what are the differences.

That's right. At least `layout order` order can be changed by `circle2circle`. `circle-editor` instead of `circle-resizer` was just a proposition without any real case. Agree that's maybe it is not a good idea. Let's keep `circle-resizer` (at least now).

> Our individual tools are integrated through one-cmds, onecc tool and also through sub tools,
if the feature is for users, to manipulate model files.
Please integrate it though this tool, and also prepare some UI tests like others,
though your draft PR, so that reviewers can provide some comments if something goes in different direction.

I see. I'll introduce such integration to draft PR but still I would like to keep Python API (built via setup.py) (in parallel to `one-cmds`). The first user (in my plans) of such API will be `circle-exir`.

> please preapre a res/TensorFlowLiteRecipes recipes with simplified version of the model or part of the model; each dim < 10 in the shape, so that luci-interpreter execution doesn't take too much time.

I see. I'll add such tests.

Thank you for reviewing the proposition.

**Comment by seanshpark** [comment]
> Do you think that resizing feature such be an option of circle2circle? 

>  I'll introduce such integration to draft PR but still I would like to keep Python API (built via setup.py) (in parallel to one-cmds). The first user (in my plans) of such API will be circle-exir.

These two seems to be linked.
So you are thinking of providing this tool for `circle-exir`, and not for `one-cmds` ? right?

**Comment by seanshpark** [comment]
> ... it doesn't covert cases fulfilled by  ...
> Do you think that resizing feature such be an option of circle2circle?

As you know, current functionality may not satisfy your requirements but it is a duplicate. For users and other developers, this may give confusion. "What should I use? why are there two? where should I contribute? ..."

Anyway before we discuss about this, IMHO, we may need to define about how we provide `circle-exir` to the users and what use-cases it will solve.

**Comment by jinevening** [comment]
Let me share my expectation. Here is the overall flow.

Torch module + user parameters -> `circle-exir` -> circle model (A) -> `circle-resizer` -> circle model (B)

1. Based on user parameters, `circle-exir` will generate circle model (A) from Torch module. circle model (A) can be either dynamic or static. (Let's pass discussion on how users will use `circle-exir` (cmdline or library). That's out of the scope of this issue.)
2. `circle-resizer` will convert circle model (A) to circle model (B). Here are some use cases in terms of LLM.
- If circle model (A) has a static seq length, e.g., 256, we can change it to 512 or 1k to support longer sequences.
- We can use chunk-sharing (https://arxiv.org/abs/2407.05858) to support dynamic sequence length. For that it would be possible to change seq length to the size of a chunk.
- For VLM, the length of image tokens can vary based on image resolution.
- ...

Actually, all of the use cases are possibly covered by changing user parameters given to `circle-exir`. Then why we need `circle-resizer`?

Sometimes running `circle-exir` multiple times for various shapes can be burdensome. For example,
- If we use pt2e quantizer in `circle-exir` and the quantizer uses a long-running algorithm, it would be not feasible to run `circle-exir` frequently.
- If `circle-exir` was used in the GPU server (possibly cloud), using the GPU server itself can be a cost (limited access, limited GPU hour, limited resources ..).
- If external users (ex: app part) have generated circle model (A) using `circle-exir`, then we have to request them to change user parameters, which takes time.

`circle-resizer` may be useful for such cases, as it allows to change model shape without running `circle-exir`.

**Comment by seanshpark** [comment]
>  circle model (A) -> circle-resizer -> circle model (B)

Q) are circle model A and B quantized?

**Comment by jinevening** [comment]
> Q) are circle model A and B quantized?

Not always, but I think so in most cases.

**Comment by seanshpark** [comment]
> I think so in most cases.

OK, I'll consider this sitation for changes.

**Comment by mbencer** [comment]
> These two seems to be linked.
So you are thinking of providing this tool for circle-exir, and not for one-cmds ? right?

I would like to provide both Python API (built via `setup.py`) and integration with `one-cmds`.
I would like to just avoid another python tool which wraps some something from `ONE/compiler` **by calling sub-process**.
Arguments: it is not easy to debug, setting some env variables is needed.

And about relation `circle-resizer` with `circle-exir`: At least now I would like to use `circle-resizer` only as a dependency to `circle-exir` **tests** to enable numerical results comparison (not to the application itself).

**Comment by seanshpark** [comment]
> I would like to provide both Python API (built via setup.py) and integration with one-cmds.

which setup.py ?

**Comment by mbencer** [comment]
> I would like to just avoid another python tool which wraps some something from ONE/compiler by calling sub-process.
    Arguments: it is not easy to debug, setting some env variables is needed.
I can't cath what you want to say. Can you plz elaborate this?

I mean about tools like `circle-interpreter` (from `python-tools` private repo) which call `onecc` as a `subprocess`.


> And about relation circle-resizer with circle-exir: At least now I would like to use circle-resizer only as a dependency to circle-exir tests to enable numerical results comparison (not to the application itself).
this too. plz add some details about the usecase you are thinking of.

For for clarification - the reasons of `circle-resizer` implementation are excellent described in https://github.com/Samsung/ONE/issues/14791#issuecomment-2756289136. In the E2E pipeline resizing will be just a separate tests.
I've mentioned about `circle-exir unit tests` just to show other dependencies. In the current moment (without `circle-resizer`) I am able to test only correctness of produced graphs based on topological comparison and calculated shapes.
With `circle-resizer` after shape specialization (from dynamic to static) I will be able to inference such test model and check numerical results.

**Comment by seanshpark** [comment]
> ... implementation are excellent described in https://github.com/Samsung/ONE/issues/14791#issuecomment-2756289136. In the E2E pipeline resizing will be just a separate tests.

using word `tests` gives confusion. As I understand, @jinevening usecase is compilation process, not a test of somekind.
what kind of `test` are you mentioning?

**Comment by mbencer** [comment]
> using word tests gives confusion. As I understand, @jinevening usecase is compilation process, not a test of somekind.
what kind of test are you mentioning?


sorry, my bad - should be `a separate step`. Too much thinking about tests 😉

**Comment by seanshpark** [comment]
@mbencer https://github.com/Samsung/ONE/issues/14791#issuecomment-2757050210

**Comment by mbencer** [comment]
> which setup.py ?

In my current version a separate `setup.py` for `circle-resizer`. The test version (dependent libs names are hardcoded) you can see in my draft - https://github.com/Samsung/ONE/pull/14727/files#diff-01f326e0963fcf9a87bd48e811343d30898dc2e182955bb6300f469ae708e35d

**Comment by mbencer** [comment]
> In my current version a separate setup.py for circle-resizer.

@seanshpark @jinevening 
In general I see that a separate wheel for only a one tool (one step from the whole compilation pipeline) is not very legible solution and can be confusing for a user.
What do you think about such unified wheel-based API for the whole compiler part (at least import, quantize, resize, optimize etc.)? As I see runtime part have something like that.
The main advantage I see is no need to save/read from/to file after each stage (the processing can be done in-memory).
Of course it's the topic for a separate issue.

**Comment by mbencer** [comment]
@seanshpark I've collected requested changes to the draft:

> Our individual tools are integrated through one-cmds, onecc tool and also through sub tools,
> if the feature is for users, to manipulate model files.
- [one-resize](https://github.com/Samsung/ONE/pull/14727/files#diff-90bed8ca2c106ccf9ca89062ead31b7a640dd698da67840e044de96a4e68f044) + tests
- [integration with one-build](https://github.com/Samsung/ONE/pull/14727/files#diff-0a7494de231bbe1eb74cb4904b0553f90688953b064dbbf8f3ef3ecebf94fcbb) + [test](https://github.com/Samsung/ONE/pull/14727/files#diff-0d5a118dbdc8f6fb1fe50a04052f799da92a70a4a05e5b9652ae493d93dd392c)
- [running via onecc](https://github.com/Samsung/ONE/pull/14727/files#diff-8b47ede7934b761817bc7d90623188bcfb1a3c0cb04c732da82367256e575623)

> Please integrate it though this tool, and also prepare some UI tests like others

Q1. The you mean a separate test binary (like `circle-resizer-tests`)?

> please preapre a res/TensorFlowLiteRecipes recipes with simplified version of the model or part of the model; each dim < 10 in the > shape, so that luci-interpreter execution doesn't take too much time.

Dynamic shapes tests based on `res/TensorFlowLiteRecipes` are pretty problematic since they are based on `shape_signature` which seems to strictly tf feature. I've pushed cut fragments of dynamic LLama (query part + normalize part in this moment) in .circle format - https://github.com/Samsung/ONE/pull/14727/files#diff-260227deca0ffece50b29c1818d35063c8b8798a5ff22ad3bb224e4b1041e33b and https://github.com/Samsung/ONE/pull/14727/files#diff-3a7a8ef5b6142928082940b155b60fa8b60795108ad25532261731a5fa35a003.
Q2. Are you ok with such approach?

About Python Package - at least in the first version I am ok with exposing only one-cmd-based API (qunanizer in `circle-exir` is also used in this way now). So maybe let's skip this part for now ;-)

In the meantime I've pushed [the first PR](https://github.com/Samsung/ONE/pull/15062).
